### PR TITLE
Fix Freeing flow - GC and timers - [MOD-6276]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

In an attempt to fix a flaky test in #4200, we found an edge case where we drop an index when the GC is about to finish its periodic run.
In this case, the GC gets to free the spec when it's the last one to have a strong reference to it, and as part of freeing resources, it also attempts to stop the spec's GC (itself). It doesn't have a timer yet so it can't stop any timer, and therefore it pushes a freeing task to the GCs' job queue. It then finishes the periodic call and **schedules another periodic for itself**.
We now have two jobs related to this GC: the first is to destroy it, and the second is to execute it, and they will be fired in this order, causing a use-after-free.

The solution is not to attempt to stop the GC ever, and always let it free itself from its periodic run when it discovers the index was dropped.

Additional fixes:

1. More timer-API-related fixes when freeing a temporary index
2. update global statistics from the main thread only
3. run vecsim tiered index GC on the current spec field, in case a new index was added using `FT.ALTER`

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/module.c [hunk 1]
```diff
@@ -660,7 +660,6 @@ static int AlterIndexInternalCommand(RedisModuleCtx *ctx, RedisModuleString **ar
     return QueryError_ReplyAndClear(ctx, &status);
   }
 
-  FieldsGlobalStats_UpdateStats(sp->fields + (sp->numFields - 1), 1);
```
Comment (MeirShpilraien) [inline src/module.c:663]: Why removing this?

File: src/spec.c [hunk 2]
```diff
@@ -277,19 +244,22 @@ static void IndexSpec_TimedOutProc(RedisModuleCtx *ctx, WeakRef w_ref) {
     // the spec was already deleted, nothing to do here
     return;
   }
-#ifdef _DEBUG
-  RedisModule_Log(NULL, "notice", "Freeing index %s by timer", sp->name);
-#endif
+  RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_VERBOSE, "Freeing index %s by timer", sp->name);
```
Comment (MeirShpilraien) [inline src/spec.c:247]: In general this is less efficient because Redis compares strings in order to decide the log level. Not sure its critical.

File: src/spec.c [hunk 3]
```diff
@@ -1063,6 +1033,10 @@ static int IndexSpec_AddFieldsInternal(IndexSpec *sp, StrongRef spec_ref, ArgsCu
   IndexSpecCache_Decref(sp->spcache);
   sp->spcache = IndexSpec_BuildSpecCache(sp);
 
+  for (size_t ii = prevNumFields; ii < sp->numFields; ++ii) {
+    FieldsGlobalStats_UpdateStats(sp->fields + ii, 1);
+  }
```
Comment (MeirShpilraien) [inline src/spec.c:1038]: Please explain why it moved here?

File: src/spec.c [hunk 4]
```diff
@@ -1181,10 +1155,6 @@ StrongRef IndexSpec_Parse(const char *name, const char **argv, int argc, QueryEr
     SchemaRule_FilterFields(spec);
   }
 
-  for (int i = 0; i < spec->numFields; i++) {
-    FieldsGlobalStats_UpdateStats(spec->fields + i, 1);
-  }
```
Comment (MeirShpilraien) [inline src/spec.c:1186]: Why removing this?

File: src/spec.c [hunk 5]
```diff
@@ -1518,12 +1500,6 @@ StrongRef IndexSpec_LoadUnsafeEx(RedisModuleCtx *ctx, IndexLoadOptions *options)
   IndexSpec_IncreasCounter(sp);
 
   if (!RS_IsMock && (sp->flags & Index_Temporary) && !(options->flags & INDEXSPEC_LOAD_NOTIMERUPDATE)) {
-    if (sp->isTimerSet) {
-      WeakRef old_timer_ref;
-      if (RedisModule_StopTimer(RSDummyContext, sp->timerId, (void **)&old_timer_ref) == REDISMODULE_OK) {
-        WeakRef_Release(old_timer_ref);
-      }
-    }
     IndexSpec_SetTimeoutTimer(sp, StrongRef_Demote(spec_ref));
```
Comment (MeirShpilraien) [inline src/spec.c:1503]: Why `RedisModule_StopTimer` was replace with `IndexSpec_SetTimeoutTimer`?

File: src/debug_commads.c [hunk 6]
```diff
@@ -665,6 +667,27 @@ DEBUG_COMMAND(GCForceBGInvoke) {
...
+    return RedisModule_ReplyWithError(ctx, "Unknown index name");
+  }
+  // wait for the timer to be called
+  uint64_t remaining;
+  if (RedisModule_GetTimerInfo(RSDummyContext, sp->gc->timerID, &remaining, NULL) == REDISMODULE_OK) {
+    usleep(1000 * remaining);
+  }
+  // Nullify the pointer so we won't try to free it in the next step.
+  // It should free itself when the timer is called and it discovers that the index was deleted.
+  sp->gc = NULL;
```
Comment (MeirShpilraien) [inline src/debug_commads.c:686]: Why we need to Nullify it? We say that we never drop the GC when freeing the index no?

File: src/debug_commads.c [hunk 7]
```diff
@@ -665,6 +667,27 @@ DEBUG_COMMAND(GCForceBGInvoke) {
   return REDISMODULE_OK;
 }
 
+DEBUG_COMMAND(GCWaitForInvokeAndDrop) {
```
Comment (MeirShpilraien) [inline src/debug_commads.c:670]: Maybe we can just add some field for FT.INFO indicate the time for the next GC run? I wonder if its enough for the test?

**Review by MeirShpilraien — COMMENTED** [review]
:+1: 
Mostly questions :)

File: src/debug_commads.c (hunk 6 above)
Comment (alonre24) [inline src/debug_commads.c:686]: Also, explain that we are holding the GIL here, and GC won't start running until we finish here

File: src/spec.c (hunk 2 above)
Comment (GuyAv46) [reply src/spec.c:247]: Right, but I think this is fine in this case

File: src/spec.c (hunk 5 above)
Comment (GuyAv46) [reply src/spec.c:1503]: it wasn't replaced. we previously checked and canceled the timer before calling this function, but this function also deals with the timer internally (code duplication). I removed the external call because the internal one is used by other flows that call `IndexSpec_SetTimeoutTimer`.

File: src/debug_commads.c (hunk 6 above)
Comment (GuyAv46) [reply src/debug_commads.c:686]: right. It was previously needed before I removed the call to stop the GC completely. will remove and make another change requested by @alonre24.

File: src/debug_commads.c (hunk 7 above)
Comment (GuyAv46) [reply src/debug_commads.c:670]: I think it's not. First of all, by the end of this function, the index is removed from all the globals and we cannot call `FT.INFO` on it. Second, the GC runs after the index is freed and doesn't have any access to its statistics anymore.

File: src/spec.c (hunk 4 above)
Comment (GuyAv46) [reply src/spec.c:1186]: Moved into the function that actually adds the fields

File: src/module.c (hunk 1 above)
Comment (GuyAv46) [reply src/module.c:663]: Moved into the function that actually adds the fields

File: src/spec.c (hunk 3 above)
Comment (GuyAv46) [reply src/spec.c:1038]: This function is used to add new fields, both by `FT.CREATE` and `FT.ALTER`. It makes sense to update the statistics here when we know we added them correctly. We also know we hold the GIL at this moment.
The complementary function was moved into `IndexSpec_RemoveFromGlobals`, where we mark the spec as deleted and make it unreachable from the outside, and we also know we have the GIL there

File: src/debug_commads.c [hunk 8]
```diff
@@ -676,16 +676,21 @@ DEBUG_COMMAND(GCWaitForInvokeAndDrop) {
...
-  // Nullify the pointer so we won't try to free it in the next step.
-  // It should free itself when the timer is called and it discovers that the index was deleted.
-  sp->gc = NULL;
+  // Drop the index
   IndexSpec_RemoveFromGlobals(ref);
-  return RedisModule_ReplyWithSimpleString(ctx, "OK");
+  // Block the client until the GC is done.
+  // Wait for extra 1 second to give the GC a chance to free itself.
+  // Since the index is already dropped, the GC will free itself quickly and unblock the client.
+  task->bClient = RedisModule_BlockClient(ctx, GCForceInvokeReply, GCForceInvokeReplyTimeout, NULL, remaining + 1000);
```
Comment (alonre24) [inline src/debug_commads.c:692]: remove the timeout here or set to 30 seconds like in force invoke

**Review by alonre24 — COMMENTED** [review]
just add test to ft.alter

File: src/debug_commads.c [hunk 9]
```diff
@@ -665,6 +667,32 @@ DEBUG_COMMAND(GCForceBGInvoke) {
...
+  StrongRef ref = IndexSpec_LoadUnsafe(ctx, RedisModule_StringPtrLen(argv[0], NULL), 0);
+  IndexSpec *sp = StrongRef_Get(ref);
+  if (!sp) {
+    return RedisModule_ReplyWithError(ctx, "Unknown index name");
+  }
+  GCTask *task = NULL;
+  uint64_t remaining;
+  while (RedisModule_GetTimerInfo(RSDummyContext, sp->gc->timerID, &remaining, (void**)&task) != REDISMODULE_OK) {
+    // The GC is running, give it a chance to finish and schedule itself again.
+    RedisModule_ThreadSafeContextUnlock(ctx);
```
Comment (MeirShpilraien) [inline src/debug_commads.c:683]: This is supper dangerous, we should never release the GIL on the main thread not even for tests.. I wonder what kind of flakiness issues this can cause. The right thing to do is to block the client and release it when the timer fired.
I wonder if we can make it more general then drop the index after the GC fired, I wonder if there are more stuff that we would like to test on those cases (maybe alter the index? maybe drop and recreate it?). I understand that we can not perform the dropping part from the client because we want atomicity. Maybe we can pass this debug command a Lua script that we will run when the timer fire?

Comment (MeirShpilraien) [reply src/debug_commads.c:683]: Ohh this is waiting for the GC to end... Anyway the comment still hold, we should not release the GIL here but instead block the client and wake it at the proper place...

File: src/debug_commads.c [hunk 10]
```diff
@@ -665,6 +667,32 @@ DEBUG_COMMAND(GCForceBGInvoke) {
...
+    usleep(1000);
+    RedisModule_ThreadSafeContextLock(ctx);
+  }
+  // Drop the index
+  IndexSpec_RemoveFromGlobals(ref);
+  // Block the client until the GC is done.
+  // Wait for extra 1 second to give the GC a chance to free itself.
+  // Since the index is already dropped, the GC will free itself quickly and unblock the client.
+  task->bClient = RedisModule_BlockClient(ctx, GCForceInvokeReply, GCForceInvokeReplyTimeout, NULL, remaining + 1000);
```
Comment (MeirShpilraien) [inline src/debug_commads.c:692]: Why counting on timeout here at all? We have a test global timeout.

Comment (MeirShpilraien) [reply src/debug_commads.c:692]: Also what if we already have a blocked client on this GC task?

File: src/gc.c [hunk 11]
```diff
@@ -77,7 +77,15 @@ static void threadCallback(void* data) {
   }
 
   if (!ret) {
-    rm_free (task);
+    // The index was freed. There is no need to reschedule the task.
+    // We need to free the task and the GC.
+    RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_DEBUG, "GC: Self-Terminating. Index was freed.");
+    gc->callbacks.onTerm(gc->gcCtx);
+    if (bc) { // A debug command is waiting for the GC to finish
```
Comment (MeirShpilraien) [inline src/gc.c:84]: I wonder why we need the `task->debug` that we check above.. here we do not check it, maybe we do not need it and we can simply count on `bc` not set to NULL?

Comment (GuyAv46) [reply src/gc.c]: there is a non-blocking debug command for triggering the GC. we do need to know if this task is debug or not

File: src/gc.c [hunk 12]
```diff
@@ -79,18 +80,20 @@ static void threadCallback(void* data) {
...
     rm_free(gc);
     goto end;
   }
 
   RedisModule_ThreadSafeContextLock(ctx);
-  gc->timerID = scheduleNext(task);
+  if (gc->timerID) {
+    gc->timerID = scheduleNext(task);
+  } else {
+    RedisModule_Log(RSDummyContext, REDISMODULE_LOGLEVEL_DEBUG, "GC %p: Not scheduling next collection", gc);
```
Comment (alonre24) [inline src/gc.c:95]: free task

File: tests/pytests/test_vecsim.py [hunk 13]
```diff
@@ -2273,17 +2273,19 @@ def expected(cur_score_name = None):
 
 
 @skip(cluster=True, noWorkers=True)
-def test_tiered_index_gc(env):
-    fork_gc_interval_sec = '10'
-    N = 1000
-    env = Env(moduleArgs=f'WORKER_THREADS 2 MT_MODE MT_MODE_FULL FORK_GC_RUN_INTERVAL {fork_gc_interval_sec}'
+def test_tiered_index_gc():
+    N = 100
+    env = Env(moduleArgs=f'WORKER_THREADS 2 MT_MODE MT_MODE_FULL FORK_GC_RUN_INTERVAL 1000000000000'
```
Comment (alonre24) [inline tests/pytests/test_vecsim.py:2278]: use stop GC

**Review by alonre24 — DISMISSED** [review]
Bravo

File: src/debug_commads.c [hunk 14]
```diff
@@ -665,6 +665,50 @@ DEBUG_COMMAND(GCForceBGInvoke) {
...
+    return RedisModule_WrongArity(ctx);
+  }
+  StrongRef ref = IndexSpec_LoadUnsafe(ctx, RedisModule_StringPtrLen(argv[0], NULL), 0);
+  IndexSpec *sp = StrongRef_Get(ref);
+  if (!sp) {
+    return RedisModule_ReplyWithError(ctx, "Unknown index name");
+  }
+  GCTask *task;
+  if (RedisModule_StopTimer(RSDummyContext, sp->gc->timerID, (void **)&task) == REDISMODULE_OK) {
+    rm_free(task);
```
Comment (MeirShpilraien) [inline src/debug_commads.c:679]: This task might have a blocked client? If so lets unblock it with an error?

File: src/gc.c [hunk 15]
```diff
@@ -77,25 +78,29 @@ static void threadCallback(void* data) {
...
+    if (bc) RedisModule_UnblockClient(bc, NULL);
+    rm_free(task);
+    rm_free(gc);
     goto end;
   }
 
   RedisModule_ThreadSafeContextLock(ctx);
-  gc->timerID = scheduleNext(task);
+  if (gc->timerID) {
+    gc->timerID = scheduleNext(task);
```
Comment (MeirShpilraien) [inline src/gc.c:93]: Theoretically we only need to lock Redis here right? Not on the else.
Also does Redis promise that the timerID we get will never be zero?

File: src/gc.c [hunk 16]
```diff
@@ -171,6 +159,16 @@ void GCContext_ForceBGInvoke(GCContext* gc) {
...
 }
 
+static void GCContext_UnblockClient(void* data) {
+  RedisModuleBlockedClient *bc = data;
+  RedisModule_BlockedClientMeasureTimeEnd(bc);
+  RedisModule_UnblockClient(bc, NULL);
+}
+
+void GCContext_WaitForAllOperations(RedisModuleBlockedClient* bc) {
+  redisearch_thpool_add_work(gcThreadpool_g, GCContext_UnblockClient, bc, THPOOL_PRIORITY_HIGH);
```
Comment (MeirShpilraien) [inline src/gc.c:169]: Do we ever use any other priority then `THPOOL_PRIORITY_HIGH`? If so it might cause an issue no?

File: src/spec.c [hunk 17]
```diff
@@ -1379,33 +1349,28 @@ void IndexSpec_Free(IndexSpec *spec) {
...
   }
   // Stop and destroy garbage collector
-  if (spec->gc) {
-    GCContext_Stop(spec->gc);
+  // We can't free it now, because it eighter runs at the moment or has a timer set which we can't
+  // deal with without the GIL.
+  // It will free itself when it discovers that the index was freed.
+  // On the worst case, it just finishes the current run and will schedule another run soon.
+  // In this case the GC will be freed on the next run, in `forkGcRunIntervalSec` seconds.
+  if (RS_IsMock && spec->gc) {
```
Comment (MeirShpilraien) [inline src/spec.c:1365]: I do not like it that we have so many code branching just for testing. What does the mock does and what does the test check? Maybe we can avoid it?

File: tests/pytests/test_issues.py [hunk 18]
```diff
@@ -801,6 +801,21 @@ def test_mod5252(env):
...
 
+@skip(cluster=True)
+def test_mod_6276(env):
+  # Setting the gc threshold to 0 so the gc won't skip its periodic run
+  env.expect('FT.CONFIG', 'SET', 'FORK_GC_CLEAN_THRESHOLD', '0').ok()
+  # Create an index and add a document + garbage
+  env.expect('FT.CREATE', 'idx', 'SCHEMA', 't', 'TEXT').ok()
+  env.expect('HSET', 'doc', 't', 'Hello').equal(1)
+  # Actual Test
+  env.expect('FT.DEBUG', 'GC_STOP_SCHEDULE', 'idx').ok()   # Stop the gc from running uncontrollably
```
Comment (MeirShpilraien) [inline tests/pytests/test_issues.py:812]: Shouldn't we also wait for currently running GC jobs?

File: tests/pytests/test_issues.py [hunk 19]
```diff
@@ -801,6 +801,21 @@ def test_mod5252(env):
   env.assertEqual(res, [1, ['key_name', 'doc']])
 
 
+@skip(cluster=True)
+def test_mod_6276(env):
+  # Setting the gc threshold to 0 so the gc won't skip its periodic run
+  env.expect('FT.CONFIG', 'SET', 'FORK_GC_CLEAN_THRESHOLD', '0').ok()
```
Comment (MeirShpilraien) [inline tests/pytests/test_issues.py:807]: Maybe we should also configure the GC interval to lower value so the test will finish faster?

**Review by MeirShpilraien — COMMENTED** [review]
Hopefully last comments :)

File: src/debug_commads.c (hunk 14 above)
Comment (GuyAv46) [reply src/debug_commads.c:679]: The periodic task never has a blocked client. I can add a verification here but only the blocking debug command use the field and adds its own task. It also has a timeout for the blocking so it won’t get stuck
Let me know if you’d still prefer that I’ll add a verification

File: src/gc.c (hunk 15 above)
Comment (GuyAv46) [reply src/gc.c:93]: 1. Right, but also for the condition check. Seems fine to me to perform the else section locked for simplicity (we should usually satisfy the condition)
2. We seems to behave this way on the GC-start function below as an error value, so I assumed it is

File: src/gc.c (hunk 16 above)
Comment (GuyAv46) [reply src/gc.c:169]: Not with this thread pool, which has a single thread that prefers the high priority. Maybe we should add an API that doesn’t require us to specify the priority

File: src/spec.c (hunk 17 above)
Comment (GuyAv46) [reply src/spec.c:1365]: I have no idea. I’ll try to check. I see that this code is covered.
Maybe we should add a compilation flag for unit tests and just have this lines compiled-out for regular compilation?

Comment (GuyAv46) [reply src/spec.c:1365]: Anyway this is a destructor function. Not in use too often

File: tests/pytests/test_issues.py (hunk 18 above)
Comment (GuyAv46) [reply tests/pytests/test_issues.py:812]: Good idea

File: tests/pytests/test_issues.py (hunk 19 above)
Comment (GuyAv46) [reply tests/pytests/test_issues.py:807]: We don’t need to. That’s why we stop it in the first place. We trigger it immediately when we need to

File: src/gc.c (hunk 15 above)
Comment (GuyAv46) [reply src/gc.c:93]: https://github.com/RediSearch/RediSearch/pull/1154#discussion_r403673323

Comment (GuyAv46) [reply src/gc.c:93]: from redis:

```
    uint64_t expiretime = ustime()+period*1000;
    uint64_t key;

    while(1) {
        key = htonu64(expiretime);
        if (raxFind(Timers, (unsigned char*)&key,sizeof(key)) == raxNotFound) {
            raxInsert(Timers,(unsigned char*)&key,sizeof(key),timer,NULL);
            break;
        } else {
            expiretime++;
        }
    }
```
So it seems very unlikely to get 0, at least the same as `UINT64_MAX`. WDYT?

File: src/debug_commads.c (hunk 14 above)
Comment (MeirShpilraien) [reply src/debug_commads.c:679]: No, seems ok.

File: src/gc.c (hunk 16 above)
Comment (MeirShpilraien) [reply src/gc.c:169]: Probably a good idea but out of the scope of this PR I believe.

File: src/spec.c (hunk 17 above)
Comment (MeirShpilraien) [reply src/spec.c:1365]: Lets keep it for the scope of this PR, I think we should delete this entire mock if it caused to much branches from the production code ...

**Review by alonre24 — APPROVED** [review]

**Review by alonre24 — APPROVED** [review]

**Review by MeirShpilraien — DISMISSED** [review]

File: src/fork_gc.c [hunk 20]
```diff
@@ -1202,12 +1204,15 @@ static int periodicCb(RedisModuleCtx *ctx, void *privdata) {
       if (getppid() != ppid_before_fork) exit(1);
     }
 #endif
-    FGC_childScanIndexes(gc);
+    // Pass the index to the child process
```
Comment (alonre24) [inline src/fork_gc.c:1207]: Explain why we take the strong ref to the child

**Review by alonre24 — APPROVED** [review]
